### PR TITLE
Terminates pods if application is high on db error

### PR DESCRIPTION
**Pull request by zhangxin511**
## Description
1. Added counters on mongo collection level
2. Added configurable failure rate to perform automatic pods stop node process

## Breaking Changes
None

### CODE REVIEW

File: server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoErrorRetryAnalyzer.ts [hunk 1]
```diff
@@ -43,23 +43,12 @@ export class MongoErrorRetryAnalyzer {
 
 	public shouldRetry(error: Error): boolean {
 		const rule = this.getRetryRule(error);
-		const sanitizedError = {
```
Comment (zhangxin511) [inline server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoErrorRetryAnalyzer.ts:46]: The `error` from this context is already sanitized, no need to sanitize again and causing logs hard to read

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 2]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
 			}
 		}
 	}
+
+	private terminateBasedOnCounterThreshold(counters: Record<string, number>): void {
+		if (this.apiFailureRateTerminationThreshold > 1) {
+			return; // If threshold set more than 1, meaning we should ever terminate and skip followings.
```
Comment (hedasilv) [inline server/routerlicious/packages/services/src/mongodb.ts:446]: nit: did you mean "If threshold set more than 1, meaning we should **never** [...]"?

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 3]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
...
+		}
+		let totalCount = 0;
+		let totalFailedCount = 0;
+		for (const [apiName, apiCounter] of Object.entries(counters)) {
```
Comment (hedasilv) [inline server/routerlicious/packages/services/src/mongodb.ts:450]: Instead of computing the total count and total failed count every time we execute this method, would it be possible to have `InMemoryApiCounters` class extended to support this functionality? Meaning it already computes total success and failure as it counts every specific API, without the need to iterate through the whole array here.

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 4]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
...
+			totalCount += apiCounter;
```
Comment (hedasilv) [inline server/routerlicious/packages/services/src/mongodb.ts:451]: Are we counting per "api type"? For example, let's say we have APIs "a", "b", and "c", and then we have APIs "a.Failed" and "d.Failed".

Does total count represent the count of "a" plus "b" plus "c"?
And total failure count represent "a.Failed" plus "d.Failed" count?

Or, does each API have their "success" and "error" counterpart? Meaning that totalCount would be the totalCount for "a", being separate from the total count for "b"?

From the code I'm guessing we are putting all together in the same bucket, which might be fine - I just want to confirm that was the intention, and if computing the failure rate in this case is indeed accurate and representative of the problems we are trying to tackle.

In other words, if there is an API that we do not care much about, but it also has the ".Failed" suffix - could that cause pod restarts?

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 5]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
...
+			// OK, no need to terminate
+			return;
+		}
+
+		// This logic is to automate the process of terminates application if db become unfunctional, so
+		// kubernetes would automatically handle the restart process.
+		Lumberjack.warning(
+			`Failure rate ${failureRate} more than threshold ${this.apiFailureRateTerminationThreshold}`,
+		);
+		process.kill(process.pid, "SIGTERM");
```
Comment (hedasilv) [inline server/routerlicious/packages/services/src/mongodb.ts:472]: We should probably try to avoid calling process.kill() here. We should try to bubble this up to the RunService infrastructure in here:

https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L87

Reasons:
1. I think we should centralize the places in our code that can terminate it. If we start having process.kill() spread throughout the codebase, it will become hard to manage.
2. I suspect that calling process.kill() immediately after the Lumberjack statement might occasionally cause missing logs.
3. The above would cause a service restart but we likely would not have the metrics available for detecting that, since we killed the process would not emit the RunService metric (which happens at the same in the link I shared above)

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 6]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
...
+		let totalCount = 0;
+		let totalFailedCount = 0;
+		for (const [apiName, apiCounter] of Object.entries(counters)) {
+			totalCount += apiCounter;
+			if (apiName.endsWith(this.failedApiCounterSuffix)) {
+				totalFailedCount += apiCounter;
+			}
+		}
+
+		if (totalCount < 100) {
```
Comment (manishgargd) [inline server/routerlicious/packages/services/src/mongodb.ts:457]: nit: it would be nice to make 100 also configurable.

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 7]
```diff
@@ -422,6 +479,8 @@ export class MongoDb implements core.IDb {
 		private readonly retryEnabled = false,
 		private readonly telemetryEnabled = false,
 		private readonly mongoErrorRetryAnalyzer: MongoErrorRetryAnalyzer,
+		private readonly apiCounterIntervalMS: number,
+		private readonly apiFailureRateTerminationThreshold: number,
```
Comment (tianzhu007) [inline server/routerlicious/packages/services/src/mongodb.ts:483]: Where is the entry point for apiFailureRateTerminationThreshold? Do we need to change the value.yaml or fluid-configmap? Thanks.

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 8]
```diff
@@ -479,6 +541,9 @@ const DefaultKeepAliveInitialDelay = 60000;
 const DefaultSocketTimeoutMS = 100000;
 const DefaultConnectionTimeoutMS = 120000;
 const DefaultMinHeartbeatFrequencyMS = 10000;
+const DefaultApiCounterIntervalMS = 60000;
```
Comment (tianzhu007) [inline server/routerlicious/packages/services/src/mongodb.ts:544]: If we can make this number larger? Maybe 10 minutes, since otherwise, it would be a lot of refreshment for different kinds of MongoDB methods based on current design, which should be costly.

File: server/routerlicious/packages/services/src/mongodb.ts [hunk 9]
```diff
@@ -414,6 +440,37 @@ export class MongoCollection<T> implements core.ICollection<T>, core.IRetryable
 			}
 		}
 	}
+
+	private terminateBasedOnCounterThreshold(counters: Record<string, number>): void {
```
Comment (tianzhu007) [inline server/routerlicious/packages/services/src/mongodb.ts:444]: Great method for calculation!

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 5 above)
Comment (shubhi1092) [reply server/routerlicious/packages/services/src/mongodb.ts:472]: Agree with Henrique about bubbling it up to RunService such that it is caught as an exception [here](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L37). Because [SIGTERM event handler](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L50) calls runner.stop() which doesnt reject the runner promise in case of SIGTERM, and even if the RunService metric is available, it will be hard to identify that it was due to the mongo failure rate.

Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:472]: Make sense, however multiple things make me to decide to use `process.kill`:
1. `process.kill` does **NOT** trigger immediately exit (`process.exit` does on the other hand). It will handled by the [SIGTERM event handler](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L50) calls runner.stop() as specified by @shubhi1092 .
2. This layer is actually the deepest layer in our managed code. I can't figure out a decent / (or even workable) way to bubble up an exception from here all the way to reach to the top level layer of `RunService`, not to mention not a single path using this layer, almost all paths are depending on this layer. So I don't think it is possible there is a way for me to be able to bubble up all the way properly for all request paths. 

If you have better ideas how to proper bubble up this, let me know.

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 7 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:483]: I defined an default value:
```
// 1 means 100%, using 2 just for safety for incorrect calculations and meaning this feature disabled
const DefaultApiFailureRateTerminationThreshold = 2;
```

And the value is read from a config:
image

If config is missing, it will use the default value
image

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 8 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:544]: This is default and this value is configurable.

The reason why I use 1 minutes interval, is to:
1. Prevent loss metrics due to other reasons caused pod stop. There are not flush-like method for the collection so use a more frequent value benefits that
2. Detect the node issue faster. The reason is db layer is very busy, so even check every minutes, we should have a lot of numbers already. For the idle time, we have a protection of minimum failure count to consider

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 4 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:451]: image

To answer your questions, I paste the code sample for better example above:
The `callerName` is important here, it is the `a`, `b`, `c` stuff in your example.
1. So if a request is success, for example `a`, then `a` counter will increase at line 393
2. But if `a` request failed, then `a.Failed` will increase at line 396

With the background how I collect metrics, now move on to how to calculate failure rate: It will loop over and will get `a`, `a.Failed`, `b`, `c`, and `c.Failed` .
So total will be the sum of all these 5, while failed will be the two ends with `.Failed`, that make the failure rate correct.


Now, what if `callerName` contains `.Failed`, that is a valid concern, but the `callerName` is a controlled variable that only visible and manageable in this class, and they are kind of meaningful names like `MongoCollection.Find`, `MongoCollection.Update`. Therefore while your concern is valid, I think in reality it is fine.

Lastly, I place the counter part in the retry code layer here, is to simply the code, so I don't have to place measurement code everywhere. The counter will not count the retry requests therefore, but I think it should be generally accurate enough. If we ever need to change, we can move the counting part to each individual request

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 3 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:450]: I think there is no need here to complex the logic of `apiCounter`:
1. for single responsibility, `apiCounter` is just a counter.
2. If we really want to consider success/failure rate, we should probably create something like `failureRateReportableApiCounter`, I don't see there is a need for that now. If there is in future, then we can design the api/logic in future. For now, the logic here is simple enough. and the only cost here is just one more iteration. Even we implement a counter can do the calculation for us, I believe it will still need one more iteration.

With these thoughts, I don't think there is a need to overdesign here

**Review by tianzhu007 — APPROVED** [review]

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 3 above)
Comment (hedasilv) [reply server/routerlicious/packages/services/src/mongodb.ts:451]: I was talking more about speed and performance. With the implementation above, we are iterating through the list of apiNames every time the function is called, which happens through the timer. Sure, the number of `apiNames` is constance (I think), but still, we are running a for loop every time which is taking CPU cycles, while the same information could likely be obtained for free if we just count the total and totalFailed numbers every time we increment the counter.

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 5 above)
Comment (hedasilv) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: What happens if we throw an error here? Can we create the error object in a way that is not retried by `requestWithRetry()` in line 374? And if we can do that, does the error bubble up to RunService, or does it get intercepted somewhere else?

Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: It will get intercepted. Think about the exception throw here the `MongoServerSelectionTimeout` exception, it if don't get intercepted, it will ends terminates the pods already. However it is not the case for production, so the exception does intercepted by someone somewhere. For example in the meeting you found we reject after 3 retries:
image

If exception bubbles up properly, the pods would already crashed. Even if we spend efforts on bubble up, it is hard to cover all and future paths given db is the deepest layer. So using signal is actually better sollution.

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 3 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:451]: You can get for free when you get total and totalFaild, but think about how many times you need to increment the total & totalFaild? For example, in 1 minutes, we make 9k success request and 1k failed request, all for findOne.

So using current approach, I increate counters 10K times, and I loop the whole counters, which give me two result, and I add two more counters in total 2 times. So in total I increment 10K+2 times, loop 1 time for size of 2.

If I also keep a total as you suggested, I need to increment 20k times, 10K for total counter, 10k for findOn related counter. I saved the rest.

So when I write my reply here, I notice what our differences are. I want to collect individual categorized counter as well so it will help me for the future sli metrics. I think you were thinking all I need is total and total failed? If that is the case, no I want categorized, like findOne, findOneFailed, update, UpdateOneFailed sort of thing.

Comment (hedasilv) [reply server/routerlicious/packages/services/src/mongodb.ts:451]: No, I meant having a separate counter that symbolizes the total count. Looping through the whole array of 10k elements every time the timer triggers seemed inefficient to me, but I don't have a good argument against your point that we need to anyways increase the totalCount and totalFailed counters every time in my suggested approach. This was just a minor optimization solution, feel free to ignore.

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 5 above)
Comment (hedasilv) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: Thanks Xin. I would prefer to not kill the process with process.kill() here. I don't think that is the best approach, and even if we need to do more work to make sure that error bubbles up to RunService, where it would be handled properly, I think that investment is worth it for the reasons I already described above. In fact, I think the purpose of this work item was to do exactly that: find out why the exception is currently being intercepted, and clear the way so that it can finally bubble up to RunService.

If you think that will significantly delay the implementation and cause further issues to the service, then I don't want to slow you down. I will defer that decision to @manishgargd, who might have more insights than me, @shubhi1092 , who recently implemented important improvements in RunService, and to you.

Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: @hedasilv @shubhi1092 , @manishgargd  Just rephrase the nodeJS api:
> Even though the name of this function is process.kill(), it is really just a signal sender, like the kill system call. The signal sent may do something other than kill the target process.

It is simple emit an event, to have the listener `RunService` to the termination for us, which is across many many layers and almost ALL code path, it is impractical to bubble up

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 3 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:451]: Just clarify, the it is not looping whole array of 10K elements. It looking only two times for the two categories, `findOne` and `findOne.Failed` in above example

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 5 above)
Comment (shubhi1092) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: Hey @zhangxin511 , my concern with raising SIGTERM here is - the RunService metric will be logged with successful = true and no information about why RunService was emitted. In other words, you wont be able to identify that the service restarted because of high db failure rate by looking at the RunService metrics. This is because in case of sigterm, [runner.stop() will resolve the runner promise](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/lambdas-driver/src/kafka-service/runner.ts#L132) and [RunService will be completed as successful](https://github.com/microsoft/FluidFramework/blob/077e696729a3bc8abf9b5c69ed3e5fc95f52c3f0/server/routerlicious/packages/services-shared/src/runner.ts#L112). Because SIGTERM doesnt always mean an exception - eg: manually deleting a pod also triggers a sigterm.

So if your goal is to be able to identify restarts due to high db failure rate, then this will not work - you will need to throw an exception for that. But if you just want to make sure that runner stops gracefully before restarting, then this would work.

Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: @shubhi1092 I understand what you are talking about. My point here is an exception throw from db layer, will not reach to the entry point of `RunService`, it will definitely intercepted somewhere, either by code controlled by us, or code controlled by express, that server automatically return 500 to instead of stopping a running application.

Comment (shubhi1092) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: In that case, we cant use RunService metric to detect this situation, and would need to rely on the kusto logs / some other mechanism for that.

Comment (manishgargd) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: This is a good discussion. @zhangxin511 - You have a good point that the exception may be intercepted before it reaches RunService. However, this can be validated in a dev environment, and based on whether it is intercepted or not, we can take a decision.

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 4 above)
Comment (manishgargd) [reply server/routerlicious/packages/services/src/mongodb.ts:452]: @zhangxin511 - Thanks for the detailed explanation! 
For failure count, should we consider only "MongoServerSelectionError" or do all the errors?

File: server/routerlicious/packages/services/src/mongodb.ts (hunk 5 above)
Comment (zhangxin511) [reply server/routerlicious/packages/services/src/mongodb.ts:473]: @manishgargd No need to deploy on a dev cluster, I can replicate easily on local env:
I places an exception throw in get delta flow, which is one layer above db:
 
image

Then request to hit the flow, all get 500:
image

The alfred didn't stop
image


In fact, when rejecting the retry, it is actually throw an exception already, but not necessary to lead to pods stop.

NOTE, this is just one flow of get delta, consider DB layer is used hundreds of flows, how would it be possible to successfully bubble up an exception all the way up to the RunService?

**Review by shubhi1092 — APPROVED** [review]

**Review by manishgargd — APPROVED** [review]
